docs: Add "Inventing the `Service` trait" guide (#585)
tower-rs/tower: guides/inventing-service.md

```rust
impl Server {
    async fn run<F, Fut>(self, handler: F) -> Result<(), Error>
    where
        F: Fn(Request) -> Fut,
        // The response future is now allowed to fail
        Fut: Future<Output = Result<Response, Error>>,
    {
        loop {
            let mut connection = receive_connection().await?;
            let request = parse_http_request(&mut connection).await?;
            match handler(request).await {
                Ok(response) => write_http_response(connection).await?,
                Err(error) => handle_error_somehow(error),
            }
        }
    }
}
```

## Adding on more behavior

Next-up you might want to add supports for timeouts. A timeout is the max
duration `handler` is allowed to take. If it doesn't produce a response within
that amount of time an error is returned.

Your first idea might be to modify `Server` to support being configured with a
timeout which it will somehow apply to the call of the handler. However it turns out
you can actually add a timeout without modifying `Server`. Using
`tokio::time::timeout` we can make a new handler function that calls our
previous `handle_request` but with a timeout of 30 seconds:

```rust

> async fn handler_with_timout(request: Request) -> Result<Response, Error> {

Reviewer: ```suggestion
async fn handler_with_timeout(request: Request) -> Result<Response, Error> {
```

    let result = tokio::time::timeout(
        Duration::from_secs(30),
        handle_request(request)
    ).await;

    match result {
        Ok(Ok(response)) => Ok(response),
        Ok(Err(error)) => Err(error),
        Err(_timeout_elapsed) => Err(Error::timeout()),
    }
}
```

This provides a quite nice separation of concerns. We were able to add a timeout
without changing any of our existing code.

Lets add one more feature in this way. Imagine we're building a JSON API and
would therefore like a `Content-Type: application/json` header on all responses.
We can wrap `handler_with_timeout` in a similar way and modify the response like
so:

```rust

> async fn handler_with_timout_and_content_type(

Reviewer: ```suggestion
async fn handler_with_timeout_and_content_type(
```

    request: Request,
) -> Result<Response, Error> {

>     let mut response = handler_with_timout(request).await?;

Reviewer: ```suggestion
    let mut response = handler_with_timeout(request).await?;
```

    response.set_header("Content-Type", "application/json");
    Ok(response)
}
```

We now have a handler that will process an HTTP request, take no longer than 30
seconds, and always have the right `Content-Type` header all without modifying
our original `handle_request` function or `Server` struct.

Designing libraries that can be extended in this way is very powerful since it
allows users to extend things without having to wait for the library maintainers
to add support for it.

However there is one problem. Our current setup doesn't scale very well. Imagine
we have many `handle_with_*` functions that each add a little bit of behavior.
Having to hard code the chain of which intermediate handler calls which isn't
great. Our current chain is

> 1. `handler_with_timout_and_content_type` which calls

Reviewer: ```suggestion
1. `handler_with_timeout_and_content_type` which calls
```

> 2. `handler_with_timout` which calls

Reviewer: ```suggestion
2. `handler_with_timeout` which calls
```

3. `handle_request` which actually processes the request

It would be nice if we could somehow compose these three functions without
having to hard code the exact order. Something like:

```rust
let final_handler = with_content_type(with_timeout(handle_request));
```

While still being able to run our handler like before:

```rust
server.run(final_handler).await?;
```

You could imagine `with_content_type` and `with_timeout` being functions that
took an argument of type `F: Fn(Request) -> Future<Output = Response>` and

> returned a closure a la `impl Fn(Request) -> Future<Output = Result<Response,

Reviewer 2: Hey, does the word `la` is a typo?

Author: Nope:

> Borrowed from French, à la means "according to" or "in the manner of,"

https://www.dictionary.com/e/slang/a-la/

I have changed it however to "like" as that is probably more understandable.

Error>>` but all these closure types quickly become hard to deal with.

## The `Handler` trait

Lets try another approach. Rather than `Server::run` accepting a closure
(`Fn(Request) -> ...`) lets make a new trait that encapsulates the same `async
fn(Request) -> Result<Response, Error>` setup:

```rust
trait Handler {
    async fn call(&mut self, request: Request) -> Result<Response, Error>;
}
```

Having a trait like this allows us to write concrete types that implement it so
we don't have to deal with `Fn`s all the time.

However Rust currently doesn't support async traits so we have two options:

1. Make `call` return a boxed future like `Pin<Box<dyn Future<Output =
   Result<Response, Error>>>`. This is what the [async-trait] crate does.
2. Add an associated `type Future` to `Handler` so users get to pick their own
   type.

Lets go with option 2 as its the most flexible. Users who have a concrete future
type can use that without the cost of a `Box` and users who don't care can still
use `Pin<Box<...>>`.

```rust
trait Handler {
    type Future: Future<Output = Result<Response, Error>>;

>     fn call(&mut self, request: Request) -> Self::Future;

Author: I would also like to add a paragraph about why `call` taking `&mut self` is necessary. As I don't think thats entirely obvious.

}
```

This is actually _almost_ the `Service` trait defined in Tower. If you've been
able to follow along until this point you now understand most of Tower. Besides
the `Service` trait Tower also provides several utilities that implement
`Service` by wrapping some other type that also implements `Service`. Exactly
like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
similar to what we've just did.

Some example services provided by Tower:

- `Timeout` - This is pretty much identical to the timeout we have built.
- `Retry` - To automatically retry failed requests.
- `RateLimit` - To limit the number of requests a service can handle over a
  period of time.

Types like `Timeout` and `JsonContentType` are typically called "middlewares"
since they wrap another `Service` and transform the request or response in some
way. Whereas types like `RequestHandler` is typically called "leaf services"
since it sits at the leaves of a tree of nested services. The actual responses
are normally produced in leaf services and modified by middlewares.

The only thing left to talk about is "backpressure" and `poll_ready`.

> ## Backpressure

Reviewer 2: I'd recommend adding a `backpressure` explanation article here. Such as this wonder one: https://medium.com/@jayphelps/backpressure-explained-the-flow-of-data-through-software-2350b3e77ce7.

This would help a lot to those new to `backpressure`. What do you think?

Author: Good idea! I added a link in 7461ae492ddaa

Imagine you wanted to write a rate limit middleware that wraps a `Service` and
puts a limit on the maximum number of concurrent requests the underlying service
can will receive. This would be useful if you had some service that had a hard
upper limit on the amount of load it could handle.

With our current `Service` trait we don't really have a good way to implement
something like that. We could try:

```rust
impl<R, T> Service<R> for ConcurrencyLimit<T> {
    fn call(&mut self, request: R) -> Self::Future {
        // 1. Check a counter for the number of running requests
        //    currently.
        // 2. If there is capacity left send the request to `T`
        //    and increment the counter.
        // 3. If not somehow wait a bit and check the capacity
        //    again.
    }
}
```

If there is no capacity left we just have to wait a bit and then check again. So
if the service is receiving a lot of requests you can imagine waiting for a long
time.  It would also be nice if we had a way to tell the caller that there was
no more capacity left. Maybe they wanted to react to that by calling another

> service instead.

Author: I'm not quite sure this argument is strong enough. Rather than "waiting for some random amount of time" one could await on an async semaphore. However that merges "driving a service readiness" and "calling the service" so `poll_ready` is still better. I'm just not quite sure this section drives that point home.

Author: Hopefully 8283ecaec7446 makes things a bit more clear.

In a way it would be nice if `Service` had a method like this:

```rust
trait Service<R> {
    async fn ready(&mut self);
}
```

`ready` would be an async function that completes when the service has capacity
enough to receive one new request. We would then require users to first call
`service.ready().await` before doing `service.call(request).await`.

> This way way `ConcurrencyLimit` could track capacity inside `ready` and not

Reviewer 2: Duplicated word `way`?

Author: Yes! Good catch.

    ).await;

    match result {
        Ok(Ok(response)) => Ok(response),
        Ok(Err(error)) => Err(error),
        Err(_timeout_elapsed) => Err(Error::timeout()),
    }
}
```

This provides a quite nice separation of concerns. We were able to add a timeout
without changing any of our existing code.

Lets add one more feature in this way. Imagine we're building a JSON API and
would therefore like a `Content-Type: application/json` header on all responses.
We can wrap `handler_with_timeout` in a similar way and modify the response like
so:

```rust
async fn handler_with_timeout_and_content_type(
    request: Request,
) -> Result<Response, Error> {
    let mut response = handler_with_timeout(request).await?;
    response.set_header("Content-Type", "application/json");
    Ok(response)
}
```

We now have a handler that will process an HTTP request, take no longer than 30
seconds, and always have the right `Content-Type` header all without modifying
our original `handle_request` function or `Server` struct.

Designing libraries that can be extended in this way is very powerful since it
allows users to extend things without having to wait for the library maintainers

> to add support for it.

Author: I added a paragraph just below this one.

having to hard code the exact order. Something like:

```rust
let final_handler = with_content_type(with_timeout(handle_request));
```

While still being able to run our handler like before:

```rust
server.run(final_handler).await?;
```

You could imagine `with_content_type` and `with_timeout` being functions that
took an argument of type `F: Fn(Request) -> Future<Output = Response>` and
returned a closure like `impl Fn(Request) -> Future<Output = Result<Response,
Error>>` but all these closure types quickly become hard to deal with.

## The `Handler` trait

Lets try another approach. Rather than `Server::run` accepting a closure
(`Fn(Request) -> ...`) lets make a new trait that encapsulates the same `async
fn(Request) -> Result<Response, Error>` setup:

```rust
trait Handler {
    async fn call(&mut self, request: Request) -> Result<Response, Error>;
}
```

Having a trait like this allows us to write concrete types that implement it so
we don't have to deal with `Fn`s all the time.

However Rust currently doesn't support async traits so we have two options:

1. Make `call` return a boxed future like `Pin<Box<dyn Future<Output =

>    Result<Response, Error>>>`. This is what the [async-trait] crate does.

Reviewer 3: Maybe its worth adding a footnote? I think for some users it could connect the dots better than others. Unfortunately with stuff like this you will encounter lifetimes (as you show later on) so maybe its worth it to bring it up?

delegate to the inner handler and let it do its thing. We don't know what it is
we just know that it produces a `Result<Response, Error>` when its done.

This code doesn't quite compile though. We get an error like this:

```
error[E0759]: `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
   --> src/lib.rs:145:29
    |
144 |       fn call(&mut self, request: Request) -> Self::Future {
    |               --------- this data with an anonymous lifetime `'_`...
145 |           Box::pin(async move {
    |  _____________________________^
146 | |             let result = tokio::time::timeout(
147 | |                 self.duration,
148 | |                 self.inner_handler.call(request),
...   |
155 | |             }
156 | |         })
    | |_________^ ...is captured here, requiring it to live as long as `'static`
```

The issue is that we're capturing a `&mut self` and moving it into an async
block. That means the lifetime of our future is tied to the lifetime of `&mut
self`. This doesn't work for us since we might want to run our response futures
on multiple threads to get better performance or produce multiple response

> futures and run them all in parallel. That isn't possible if a reference to the

Reviewer 3: This isn't actually the reason for this error, basically, TLDR we don't have GAT. All associated types must return a `'static` reference and since `'static` > `'a`  from `&mut 'a self` it doesn't like it. Async-trait gets around this above as I mentioned in that comment, plus the fact that it is a trait object so a concrete type vs a generic like we have here.

Author: Yes thats totally correct. Again I'm not sure how much lifetime trouble to actually go into. I'm a bit worried that if we say `Pin<Box<dyn Future>>` needs a `+ 'a` from `&'a mut self` then some users might wonder "well easy, just define service as `trait Service<'a>`". For example this does compile:

```rust
use std::future::Future;

trait Service<'a, Request> {
    type Response;
    type Error;
    type Future: Future<Output = Result<Self::Response, Self::Error>> + 'a;

    fn call(&'a mut self, request: Request) -> Self::Future;
}

struct HttpRequest;
struct Error;

async fn call_all<S>(mut svc: S, requests: Vec<HttpRequest>) -> Result<(), Error>
where
    for<'a> S: Service<'a, HttpRequest, Error = Error>,
{
    for request in requests {
        svc.call(request).await?;
    }
    
    Ok(())
}
```

But we can't `tokio::spawn` the response as that requires the future to be `'static` which then makes `self` `&'static mut self` which clearly is quite limiting.

So yeah you're totally correct but I'm trying to say something that will make users think "yeah that kinda makes sense" without opening a can of worms 😅

What do you think?

Reviewer 3: I think you're right there is a balance to be achieved here 😄

Reviewer 4: I think it's worth leaving a note here saying that `clone` isn't the only way to do this, and that in the future we're hoping to be able to express this in an even cleaner way (link to GAT issue). That's probably enough for the casual reader.

Author: Adding it as a footnote is a good idea!

handler lives inside the futures.

> Instead we need convert the `&mut self` into an owned `self`. That is exactly

Reviewer 3: another option is because the inner `Handle::Future` _has_ to be `'static` you can:

```
let fut = self.inner.call(...);
```

and that `fut` has to be a `'static` future.

Reviewer 3: The clone method also works though :D

                this.inner_handler.call(request),
            ).await;

            match result {
                Ok(Ok(response)) => Ok(response),
                Ok(Err(error)) => Err(error),
                Err(_timeout) => Err(Error::timeout()),
            }
        })
    }
}
```

Note that cloning is very cheap in this case since `RequestHandler` doesn't have
any data and `Timeout<T>` only adds a `Duration`.

One step closer. We now get a different error:

```
error[E0310]: the parameter type `T` may not live long enough
   --> src/lib.rs:149:9
    |
140 |   impl<T> Handler for Timeout<T>
    |        - help: consider adding an explicit lifetime bound...: `T: 'static`
...
149 | /         Box::pin(async move {
150 | |             let result = tokio::time::timeout(
151 | |                 this.duration,
152 | |                 this.inner_handler.call(request),
...   |
159 | |             }
160 | |         })
    | |__________^ ...so that the type `impl Future` will meet its required lifetime bounds
```

The problem now is that `T` can be any type whatsoever. It can even be a type
that contains references like `Vec<&'a str>`. However that wont work for the

> same reason as before. We need the response future to have a `'static` lifetime

Reviewer 3: This is because again our associated type is `Pin<Box<dyn Future + 'static>>` (if you don't include a lifetime its auto a `'static`. And again the same problem as above.

```rust
JsonContentType {
    inner_handler: Timeout {
        inner_handler: RequestHandler,
        duration: Duration::from_secs(30),
    },
}
```

And if we add some `new` methods to our types they become a little easier to
compose:

```rust
let handler = RequestHandler;
let handler = Timeout::new(handler, Duration::from_secs(30));
let handler = JsonContentType::new(handler);

// `handler` has type `JsonContentType<Timeout<RequestHandler>>`

server.run(handler).await
```

This works quite well! We're now able to layer on additional functionality to
`RequestHandler` without having to modify its implementation. In theory we could
put our `JsonContentType` and `Timeout` handlers into a crate and release it as
library on crates.io for other people to use!

## Making `Handler` more flexible

Our little `Handler` trait is working quite nicely but currently it only
supports our `Request` and `Response` types. It would be nice if those where
generic so users could use whatever type they want:

```rust
// Lets make the request type a generic type parameter.
// This allows users to implement `Handler` multiple times for
// the same type. Its not something we need but our users might.

> trait Handler<Request> {

Reviewer 3: Might want to add a note why the `Request` is generic but `Response` is an associated type. This is because generally for traits if its a param for method the generic is much easier while a return type associated types work much better with how you compose things.

Reviewer 4: Here's how I'd phrase this:

> We make `Request` a generic type parameter of the trait so that a given service can accept many different types of requests. We make `Response` an associated type because for any _given_ `Request` type, there can only be one (associated) `Response` type: the one the corresponding `call` returns!

Author: Love it!

    // `Request` type paramter and not the `Request` type we've used
    // until now
    fn call(&mut self, request: Request) -> Self::Future;
}
```

Our implementation for `RequestHandler` now becomes

```rust
impl Handler<Request> for RequestHandler {
    type Response = Response;
    type Error = Error;
    type Future = Pin<Box<dyn Future<Output = Result<Response, Error>>>>;

    fn call(&mut self, request: Request) -> Self::Future {
        // same as before
    }
}
```

`Timeout<T>` is a bit different. Since it wraps some other `Handler` and adds an
async timeout it actually doesn't care about what the request or response types
are. As long as the `Handler` it is wrapping uses the same types.

The `Error` type is a bit different. Since `tokio::time::timeout` returns
`Result<T, tokio::time::error::Elapsed>` we must be able to convert a
`tokio::time::error::Elapsed` into the inner `Handler`'s error type.

If we put all those things together we get

```rust
// `Timeout` accepts any request of type `R` as long as `T`
// accepts the same type of request
impl<R, T> Handler<R> for Timeout<T>
where
    // The actual type of request must not contain

>     // references. The compiler would tell us to add

Reviewer 3: bless Dr compiler 🙌

Author: 🙇‍♂️

headers.

Since our `Handler` trait is useful for defining both servers and clients,
`Handler` probably isn't an appropriate name anymore. A client doesn't handle an
request, it sends it to a server and the server handles it. Lets instead call
our trait `Service`:

```rust
trait Service<Request> {
    type Response;
    type Error;
    type Future: Future<Output = Result<Self::Response, Self::Error>>;

    fn call(&mut self, request: Request) -> Self::Future;
}
```

This is actually _almost_ the `Service` trait defined in Tower. If you've been
able to follow along until this point you now understand most of Tower. Besides
the `Service` trait Tower also provides several utilities that implement
`Service` by wrapping some other type that also implements `Service`. Exactly
like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
similar to what we've just did.

Some example services provided by Tower:

- `Timeout` - This is pretty much identical to the timeout we have built.
- `Retry` - To automatically retry failed requests.
- `RateLimit` - To limit the number of requests a service can handle over a
  period of time.

> Types like `Timeout` and `JsonContentType` are typically called "middlewares"

Reviewer 3: ```suggestion
Types like `Timeout` and `JsonContentType` are typically called "middleware"
```

I don't think this needs to be plural when its in quotes like this?

since they wrap another `Service` and transform the request or response in some
way. Whereas types like `RequestHandler` is typically called "leaf services"
since it sits at the leaves of a tree of nested services. The actual responses

> are normally produced in leaf services and modified by middlewares.

Reviewer 3: I might mention something about how they are usually the things interacting with io. We use the term "leaf" futures for the futures that interact direction with `TcpStream`, etc.

The only thing left to talk about is "backpressure" and `poll_ready`.

## Backpressure

> Imagine you wanted to write a rate limit middleware that wraps a `Service` and

Reviewer 3: Might wanna link https://aws.amazon.com/builders-library/using-load-shedding-to-avoid-overload/ and there are a few other articles in there that are REALLY good and connect the real world with these concepts. Maybe for later in this section.

Reviewer 3: Yeah, there are a few articles in that builder library that are great reads and connect some real world concepts. Load shedding tends to be the root behavior where you reject requests when the buffer is full and that rejection stops requests from coming in etc. What ever you feel is correct :)

Author: I actually thinking linking to both is fine 😊

puts a limit on the maximum number of concurrent requests the underlying service

> can will receive. This would be useful if you had some service that had a hard

Reviewer 3: can will? lol

Author: I can will good English.

upper limit on the amount of load it could handle.

With our current `Service` trait we don't really have a good way to implement
something like that. We could try:

```rust
impl<R, T> Service<R> for ConcurrencyLimit<T> {
    fn call(&mut self, request: R) -> Self::Future {
        // 1. Check a counter for the number of running requests
        //    currently.
        // 2. If there is capacity left send the request to `T`
        //    and increment the counter.
        // 3. If not somehow wait until capacity becomes available.
    }
}
```

If there is no capacity left we have to wait and somehow get notified when
capacity becomes available. So if the service is receiving a lot of requests we
risk spending a lot of time waiting. It would be nice if we had a way to tell
the caller that there was no more capacity left. Maybe they wanted to react to
that by calling another service or somehow creating another service to handle

> part of the load.

Reviewer 3: or worst case, we keep accepting requests and that results in unbounded memory growth.

Author: Good point 👍 Added a note about that.

In a way it would be nice if `Service` had a method like this:

```rust
trait Service<R> {
    async fn ready(&mut self);
}
```

`ready` would be an async function that completes when the service has capacity
enough to receive one new request. We would then require users to first call
`service.ready().await` before doing `service.call(request).await`.

Separating "calling the service" from "reserving capacity" also unlocks new use
cases such as being able to maintain a set of "ready services" that we keep up
to date in the background such that when a request arrives we already have a

> ready service to send it to and don't have to first wait for it to become ready.

Reviewer 3: or if its required to do a lot of work to say allocate the request you only need to do it when you can ensure it will get processed. (Think like reading a big image off the wire etc)

With this setup `ConcurrencyLimit` could track capacity inside `ready` and not
allow users to call `call` until there is sufficient capacity.

`Service`s that don't care about capacity can just return immediately from
`ready` or delegate to some inner `Service` they're wrapping.

However we still cannot define async functions inside traits. We could add
another associated type to `Service` called `ReadyFuture` but having to return a
`Future` will give us the same lifetime issues we've run into before. It would
be nice if there was someway around that.

> Instead we can take some inspiration from the `Future` trait and define a method

Reviewer 3: actually technically the inspiration came from `Sink` since it originally had a `start_send` but I think @carllerche wrote that too?

Reviewer 3: Yeah totally fair, just thought I'd mention it xD

called.

It would still be possible get a `Future` that resolves when capacity is
available using something like [`futures::future::poll_fn`] (or
[`tower::ServiceExt::ready`]).

This concept of services communicating with their callers about their capacity
is called "backpressure propagation". The fundamental idea is that you shouldn't
send a request to a service that doesn't have the capacity to handle it. Instead
you should wait, call another service, or somehow scale up the number of
services available. You can learn more about the general concept of backpressure
[here][backpressure].

Finally, it might also be possible for some error to happen while reserving
capacity so `poll_ready` probably should return `Poll<Result<(), Self::Error>>`.
With this change we've now arrived at the complete `tower::Service` trait:

```rust
pub trait Service<Request> {
    type Response;
    type Error;
    type Future: Future<Output = Result<Self::Response, Self::Error>>;

    fn poll_ready(
        &mut self,
        cx: &mut Context<'_>,
    ) -> Poll<Result<(), Self::Error>>;

    fn call(&mut self, req: Request) -> Self::Future;
}
```

> Middlewares that care about backpressure is pretty rare but it does enable some

Reviewer 3: buffering is the other big one! This is what like TCP does with its kernel buffer.

Author: I'm not familiar with that. Would you mention that, and if so how would you phrase it?

since it sits at the leaves of a tree of nested services. The actual responses
are normally produced in leaf services and modified by middlewares.

The only thing left to talk about is "backpressure" and `poll_ready`.

## Backpressure

Imagine you wanted to write a rate limit middleware that wraps a `Service` and
puts a limit on the maximum number of concurrent requests the underlying service
will receive. This would be useful if you had some service that had a hard upper
limit on the amount of load it could handle.

With our current `Service` trait we don't really have a good way to implement
something like that. We could try:

```rust
impl<R, T> Service<R> for ConcurrencyLimit<T> {
    fn call(&mut self, request: R) -> Self::Future {
        // 1. Check a counter for the number of running requests
        //    currently.
        // 2. If there is capacity left send the request to `T`
        //    and increment the counter.
        // 3. If not somehow wait until capacity becomes available.
    }
}
```

If there is no capacity left we have to wait and somehow get notified when
capacity becomes available. So if the service is receiving a lot of requests we

> risk spending a lot of time waiting. It would be nice if we had a way to tell

Reviewer 4: "spending a lot of time waiting" struck me as a somewhat odd way of expressing the problem. After all, the future will just be asleep, so what's the issue? You later get onto what I think is the _real_ problem, namely that you have to already have the request in memory in order to invoke `call`, and therefore you can only exert backpressure after you've already accepted a request, which leads to buffering up a lot of requests internally in the server. I think it's probably worth hoisting that point up here directly.

Author: Thats a very good point! I'll fix that :)

```rust
JsonContentType {
    inner_handler: Timeout {
        inner_handler: RequestHandler,
        duration: Duration::from_secs(30),
    },
}
```

And if we add some `new` methods to our types they become a little easier to
compose:

```rust
let handler = RequestHandler;
let handler = Timeout::new(handler, Duration::from_secs(30));
let handler = JsonContentType::new(handler);

// `handler` has type `JsonContentType<Timeout<RequestHandler>>`

server.run(handler).await
```

This works quite well! We're now able to layer on additional functionality to
`RequestHandler` without having to modify its implementation. In theory we could
put our `JsonContentType` and `Timeout` handlers into a crate and release it as
library on crates.io for other people to use!

## Making `Handler` more flexible

Our little `Handler` trait is working quite nicely but currently it only
supports our `Request` and `Response` types. It would be nice if those where
generic so users could use whatever type they want.

We make `Request` a generic type parameter of the trait so that a given service
can accept many different types of requests. We make `Response` an associated type

> because for any given `Request` type, there can only be one (associated) `Response`

Reviewer 4: ```suggestion
because for any _given_ `Request` type, there can only be one (associated) `Response`
```

    type Response;
    type Error;
    type Future: Future<Output = Result<Self::Response, Self::Error>>;

    fn call(&mut self, request: Request) -> Self::Future;
}
```

This is actually _almost_ the `Service` trait defined in Tower. If you've been
able to follow along until this point you now understand most of Tower. Besides
the `Service` trait Tower also provides several utilities that implement
`Service` by wrapping some other type that also implements `Service`. Exactly
like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
similar to what we've done thus far.

Some example services provided by Tower:

- `Timeout` - This is pretty much identical to the timeout we have built.
- `Retry` - To automatically retry failed requests.
- `RateLimit` - To limit the number of requests a service can handle over a
  period of time.

Types like `Timeout` and `JsonContentType` are typically called "middleware"
since they wrap another `Service` and transform the request or response in some
way. Whereas types like `RequestHandler` is typically called a "leaf service"
since it sits at the leaves of a tree of nested services. The actual responses

> are normally produced in leaf services and modified by middlewares.

Reviewer 5: maybe also reference https://docs.rs/tower/0.4.7/tower/trait.Service.html#middleware--layer here?

The only thing left to talk about is "backpressure" and `poll_ready`.

## Backpressure

Imagine you wanted to write a rate limit middleware that wraps a `Service` and
puts a limit on the maximum number of concurrent requests the underlying service
will receive. This would be useful if you had some service that had a hard upper
limit on the amount of load it could handle.

> With our current `Service` trait we don't really have a good way to implement

Reviewer 5: ```suggestion
With our current `Service` trait, we don't really have a good way to implement
```

something like that. We could try:

```rust
impl<R, T> Service<R> for ConcurrencyLimit<T> {
    fn call(&mut self, request: R) -> Self::Future {
        // 1. Check a counter for the number of requests currently being
        //    processed.
        // 2. If there is capacity left send the request to `T`
        //    and increment the counter.
        // 3. If not somehow wait until capacity becomes available.
        // 4. When the response has been produced, decrement the counter.
    }
}
```

> If there is no capacity left we have to wait and somehow get notified when
> capacity becomes available. Additionally we have to keep the request in memory
> while we're waiting (also called buffering). It would be more robust to only
> allocate space for the request when we are sure the service has capacity to
> handle it. Otherwise we risk using a lot of memory buffering requests while we
> wait for our service to become ready.

Reviewer 5: ```suggestion
If there is no capacity left, we have to wait and somehow get notified when
capacity becomes available. Additionally, we have to keep the request in memory
while we're waiting (also called _buffering_). This means that the more requests
that are waiting for capacity, the more memory our program would use --- if more 
requests are produced than our service can handle, we might run out of memory! 
It would be more robust to only allocate space for the request when we are sure
the service has capacity to handle it. Otherwise, we risk using a lot of memory 
buffering requests while we wait for our service to become ready.
```

> In a way it would be nice if `Service` had a method like this:

Reviewer 5: ```suggestion
It would be nice if `Service` had a method like this:
```

```rust
trait Service<R> {
    async fn ready(&mut self);
}
```

`ready` would be an async function that completes when the service has capacity
enough to receive one new request. We would then require users to first call
`service.ready().await` before doing `service.call(request).await`.

Separating "calling the service" from "reserving capacity" also unlocks new use

> cases such as being able to maintain a set of "ready services" that we keep up

Reviewer 5: ```suggestion
cases, such as being able to maintain a set of "ready services" that we keep up
```

> to date in the background such that when a request arrives we already have a

Reviewer 5: ```suggestion
to date in the background, so that when a request arrives we already have a
```

ready service to send it to and don't have to first wait for it to become ready.

> With this setup `ConcurrencyLimit` could track capacity inside `ready` and not

Reviewer 5: ```suggestion
With this design, `ConcurrencyLimit` could track capacity inside `ready` and not
```

allow users to call `call` until there is sufficient capacity.

`Service`s that don't care about capacity can just return immediately from

> `ready` or delegate to some inner `Service` they're wrapping.

Reviewer 5: ```suggestion
`ready`, or if they wrap some inner `Service`, they could delegate to its
`ready` method.
```

> However we still cannot define async functions inside traits. We could add
> another associated type to `Service` called `ReadyFuture` but having to return a
> `Future` will give us the same lifetime issues we've run into before. It would
> be nice if there was someway around that.

Reviewer 5: ```suggestion
However, we still cannot define async functions inside traits. We could add
another associated type to `Service` called `ReadyFuture`, but having to return a
`Future` will give us the same lifetime issues we've run into before. It would
be nice if there was some way around that.
```

> Instead we can take some inspiration from the `Future` trait and define a method

Reviewer 5: ```suggestion
Instead, we can take some inspiration from the `Future` trait and define a method
```

called `poll_ready`:

```rust
use std::task::{Context, Poll};

trait Service<R> {
    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<()>;
}
```

This means if the service is at capacity `poll_ready` will return

> `Poll::Pending` and somehow notify the caller, using the waker from the
> `Context`, when capacity becomes available. At that point `poll_ready` can be

Reviewer 4: ```suggestion
`Poll::Pending` and notify the caller using the waker from the
`Context` when capacity becomes available. At that point `poll_ready` can be
```

> This means if the service is at capacity `poll_ready` will return
> `Poll::Pending` and notify the caller using the waker from the
> `Context` when capacity becomes available. At that point `poll_ready` can be
> called again and if it returns `Poll::Ready(())` then capacity would be reserved
> and `call` can be called.

Reviewer 5: ```suggestion
This means if the service is at capacity, `poll_ready` will return
`Poll::Pending` and notify the caller using the waker from the
`Context` when capacity becomes available. At that point, `poll_ready` can be
called again, and if it returns `Poll::Ready(())` then capacity is  reserved and
`call` can be called.
```

Note that there is technically nothing that prevents users from calling `call`

> without first making sure the service is ready. Doing so however is considered a

Reviewer 5: ```suggestion
without first making sure the service is ready. However, doing so is considered a
```

> violation of the `Service` contract and implementations are allowed to `panic`

Reviewer 5: ```suggestion
violation of the `Service` API contract, and implementations are allowed to `panic`
```

> if `call` is called on a service that isn't ready!

Reviewer 4: Should we mention here that getting `Ready` means a slot is "reserved" for us, and that we should then use it or we're holding up capacity in the system? Plus maybe a footnote with a link to the `disarm` stuff?

`poll_ready` not returning a `Future` also means we're able to quickly check if
a service is ready without being forced to wait for it to become ready. If we

> call `poll_ready` and get back `Poll::Pending` we can simply decide to do

Reviewer 5: ```suggestion
call `poll_ready` and get back `Poll::Pending`, we can simply decide to do
```

something else instead of waiting. Among other things, this allows you to build

> load balancers that estimates the load of services by how often they return

Reviewer 5: ```suggestion
load balancers that estimate the load of services by how often they return
```

> `Poll::Pending` and sends requests to the service with the least load.

Reviewer 5: ```suggestion
`Poll::Pending`, and send requests to the service with the least load.
```

It would still be possible get a `Future` that resolves when capacity is
available using something like [`futures::future::poll_fn`] (or
[`tower::ServiceExt::ready`]).

This concept of services communicating with their callers about their capacity

> is called "backpressure propagation". Think of it as services pushing back on

Reviewer 5: maybe:
```suggestion
is called "backpressure propagation". You can think of it as services pushing back on
```

> their callers telling them to slow down if they're going too fast. The

Reviewer 5: ```suggestion
their callers, and telling them to slow down if they're producing requests too fast. The
```

fundamental idea is that you shouldn't send a request to a service that doesn't
have the capacity to handle it. Instead you should wait (buffering), drop the
request (load shedding), or handle the lack of capacity in some other way. You
can learn more about the general concept of backpressure [here][backpressure]
and [here][backpressure2].

Finally, it might also be possible for some error to happen while reserving

> capacity so `poll_ready` probably should return `Poll<Result<(), Self::Error>>`.

Reviewer 5: ```suggestion
capacity, so `poll_ready` probably should return `Poll<Result<(), Self::Error>>`.
```

With this change we've now arrived at the complete `tower::Service` trait:

```rust
pub trait Service<Request> {
    type Response;
    type Error;
    type Future: Future<Output = Result<Self::Response, Self::Error>>;

    fn poll_ready(
        &mut self,
        cx: &mut Context<'_>,
    ) -> Poll<Result<(), Self::Error>>;

    fn call(&mut self, req: Request) -> Self::Future;
}
```

> Middlewares that care about backpressure are pretty rare but it does enable some
> interesting use cases such as various kinds of rate limiting, load balancing,
> and auto scaling.

Reviewer 5: maybe
```suggestion
Many middleware don't add their own backpressure, and simply delegate to the
wrapped service's `poll_ready_. However, backpressure in middleware does enable
some interesting use cases, such as various kinds of rate limiting, load balancing,
and auto scaling.
```

> Since you never know exactly which middlewares a `Service` might consist of it

Reviewer 5: ```suggestion
Since you never know exactly which middleware a `Service` might consist of, it
```

is important to not forget about `poll_ready`.

With all this in place, the most common way to call a service is:

```rust
use tower::{
    Service,
    // for the `ready` method
    ServiceExt,
};

let response = service
    // wait for the service to have capacity
    .ready().await?
    // send the request
    .call(request).await?;
```

---

## Footnotes

<a name="pin">1</a>: There has been some discussion around whether `call` should

> take `Pin<&mut Self>` or not but so far we've decided to go with a plain `&mut

Reviewer 5: ```suggestion
take `Pin<&mut Self>` or not, but so far we've decided to go with a plain `&mut
```

> self` which means handlers (ahem services) must be `Unpin`. In practice that is

Reviewer 5: ```suggestion
self` which means handlers (ahem, _services_) must be `Unpin`. In practice that is
```

rarely an issue. More details [here](https://github.com/tower-rs/tower/issues/319).

<a name="gats">2</a>: To be a bit more precise, the reason this requires the
response future to be `'static` is that writing `Box<dyn Future>` actually

> becomes `Box<dyn Future + 'static>` which the anonymous lifetime in `fn call(&'_

Reviewer 5: ```suggestion
becomes `Box<dyn Future + 'static>`, which the anonymous lifetime in `fn call(&'_
```

> mut self, ...)` doesn't satisfy. In the future we're hoping to use [Generic
> associated types][gat] get around this. That will allow us to define the
> response future as `type Future<'a>` and `call` as `fn call<'a>(&'a mut self,
> ...) -> Self::Future<'a>`.

Reviewer 5: ```suggestion
mut self, ...)` doesn't satisfy. In the future, the Rust compiler team plans to
a feature called [generic associated types][gat] which will let us get around this.
Generic associated types will allow us to define the response future as
`type Future<'a>`, and `call` as `fn call<'a>(&'a mut self, ...) -> Self::Future<'a>`.
```
